Firestore/core/src/firebase/firestore/util/string_apple.cc

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/util/string_apple.h"
+
+#include <string>
+
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
```

Comment:
My knowledge of Objective-C is poor, so this might be totally inapplicable, but just to check -- there appear to be several methods in `NSString` that are roughly equivalent to `UTF8String`. Do all of them have the same performance issue as `UTF8String`?

1. `std::string result {[ns_str cStringUsingEncoding:NSUTF8StringEncoding], ns_str.length};`
This seems least likely to work, because `UTF8String` is probably just a shorthand for this method.

2. `std::string result{ns_str.length + 1, '\0'}; [ns_str getCString:&result[0] maxLength: ns_str.length + 1 encoding:NSUTF8StringEncoding]};`
This looks like it might be better because it might avoid allocating an `NSData` object.

3. [getBytes...](https://developer.apple.com/documentation/foundation/nsstring/1413453-getbytes?language=objc) -- probably equivalent to 2), but who knows.

Reply:
Part of the motivation for this is that we can use `CFStringRef` in straight C and C++ code without the need to compile in a separate .m or .mm file. `CFStringRef` and `NSString*` are essentially aliases and can be exchanged without any cost. Making this code work on `CFStringRef` makes it possible to perform this conversion for both types.